### PR TITLE
Added constant support to shaders

### LINKED ISSUE #10577: 3.0 Shader Language: Implement constants
i see no 'const' word anywhere, so the best way to implement const is 'uniform' but it is too heavy for such small things. i think it is important to be able to declare constants, because shader compilers can optimize shaders very well using this knowledge.

may be godot detects constants itself and outputs them as const into glsl (variables that are declared but never changed)? if so, please confirm.

### PR DESCRIPTION
**Pull request by Chaosus**
Remake of https://github.com/godotengine/godot/pull/24724 but without global variables(as @reduz requested).

Closes #10577 (only)

### CODE REVIEW

**Comment by DavidSichma** [comment]
Looks like I am just a little too late back from my Github hiatus to do that myself.

Constants are missing from completion, is there a reason for that?

If I am not mistaken there is no major change in here? Then this looks good to me.

File: drivers/gles2/shader_compiler_gles2.cpp [hunk 1]
```diff
@@ -361,6 +361,23 @@ String ShaderCompilerGLES2::_dump_node_code(SL::Node *p_node, int p_level, Gener
...
 
+			// constants
+
+			for (Map<StringName, SL::ShaderNode::Constant>::Element *E = snode->constants.front(); E; E = E->next()) {
+				String gcode;
+				gcode += "const ";
+				gcode += _prestr(E->get().precision);
+				gcode += _typestr(E->get().type);
+				gcode += " " + _mkid(E->key());
+				if (E->get().initializer != NULL) {
```
Comment (DavidSichma) [inline drivers/gles2/shader_compiler_gles2.cpp]: As constants always have to be initilized this if statement is now kinda unnecessary. Also in GLES3.

**Comment by Chaosus** [comment]
> Constants are missing from completion, is there a reason for that?

Looks like all global keywords completions currently misses out. There should be a new PR to fix it for all.

> If I am not mistaken there is no major change in here? Then this looks good to me.

It's mostly copy/paste from your PR(with few small changes). If you really want I can close this PR to present the rights to do it for you (here I've mentioned you as co-author however).

File: drivers/gles2/shader_compiler_gles2.cpp (hunk 1 above)
Comment (Chaosus) [reply drivers/gles2/shader_compiler_gles2.cpp]: Fixed

File: servers/visual/shader_language.h [hunk 2]
```diff
@@ -632,6 +641,7 @@ class ShaderLanguage {
 		IDENTIFIER_FUNCTION_ARGUMENT,
 		IDENTIFIER_LOCAL_VAR,
 		IDENTIFIER_BUILTIN_VAR,
+		IDENTIFIER_CONSTANT
```
Comment (akien-mga) [inline servers/visual/shader_language.h]: Please add trailing comma.

Comment (Chaosus) [reply servers/visual/shader_language.h]: Fixed...

**Comment by DavidSichma** [comment]
I am fine with you doing the PR.

For the completion, uniforms and varyings work for me at least. In my initial PR I did the same thing for globals. Lines: https://github.com/Chaosus/godot/blob/c2d4abf62e3673c976adfef06ef34852d4dce248/servers/visual/shader_language.cpp#L4763

**Comment by akien-mga** [comment]
Thanks to both of you! :)

**Pull request merged**